compiler/rustc_mir/src/transform/coverage/spans.rs

```diff
@@ -663,10 +671,13 @@ impl<'a, 'tcx> CoverageSpans<'a, 'tcx> {
             self.push_refined_span(pre_closure);
         }
         if has_post_closure_span {
-            // Update prev.span to start after the closure (and discard curr)
+            // Mutate `prev.span()` to start after the closure (and discard curr).
+            // (**NEVER** update `prev_original_span` because it affects the assumptions
+            // about how the `CoverageSpan`s are ordered.)
             self.prev_mut().span = self.prev().span.with_lo(right_cutoff);
-            self.prev_original_span = self.prev().span;
```

Comment:
This line updating `prev_original_span` to match the mutated `self.prev().span` has been wrong for a long time. It's clear, now, that the original_span should never be mutated, and this was always a bug.

I think recent changes elsewhere may have exposed this bug, but even so, it still occurs very rarely.

I added some tests to try to replicate the failures I saw when compiling a very large codebase, but it's very hard to replicate. (It seems to occur when calling a macro that contains a closure and returns a Result, checked with the `?` try operator. But even that is not enough to replicate the problem. It also requires that the macro code branches, and that it produces just the right order of dominator relationship between basic blocks, with just the right nested but equal spans.

The bug is clear to me even if the specific situation leading up to that bug is not as clear. I validated this against the code that triggered the error (at least 3 instances) and removing this line corrected the problem, allowing the codebase to compile to completion.